core/src/sumeragi/main_loop.rs

```diff
@@ -399,11 +399,14 @@ fn commit_block<F: FaultInjection>(
         .apply(&committed_block)
         .expect("Failed to apply block on WSV. Bailing.");
     sumeragi.send_events(state.wsv.events_buffer.replace(Vec::new()));
-    sumeragi.send_events(&committed_block);
 
     // Update WSV copy that is public facing
     *sumeragi.wsv.lock() = state.wsv.clone();
 
+    // This sends "Block committed" event, so it should be done
+    // AFTER public facing WSV update
+    sumeragi.send_events(&committed_block);
```

Comment:
considering this error was repeated twice maybe you can extract it into a method on `sumeragi`? like `update_wsv` or `finalize_block`?

Reply:
actually, everything following this line is duplicated and can be extracted into a function

Reply:
I agree, I don't like this too. But I just wanted to fix it somehow. My PR is pretty big already and I don't want to touch `Sumeragi` refactoring. I'll create an issue

Reply:
Created #3254